contracts/root/StakeManager.sol

```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+  address[] public currentValidators;
 
-  //
-  // Constructor
-  //
+  mapping (address=>Staker) stakers; 
+  mapping (uint256=>address) stakingIdToAddress;
 
-  constructor (address _token) public {
+  constructor(address _token) public payable {
     require(_token != 0x0);
     tokenObj = ERC20(_token);
+    stakeQueue = new PriorityQueue();
   }
 
-  //
-  // Modifiers
-  //
-
   // only staker
   modifier onlyStaker() {
     require(totalStakedFor(msg.sender) > 0);
     _;
   }
 
-  //
-  // ERC900 implementation
-  //
+  
+  function _priority(address user, uint256 amount, bytes data)internal view returns(uint256){
+    // priority = priority << 64 | amount.div(totalStake) ;
+    // return amount.mul(10000000).add(currentEpoch.mul(1000).add(amount.mul(100).div(user.balance)));
+    return amount;
+  }
 
   function stake(uint256 amount, bytes data) public {
-    stakeFor(msg.sender, amount, data);
+    // add condition to update already existing users stake !? 
+    if (stakers[msg.sender].epoch==0) { 
+      stakeFor(msg.sender, amount, data);
+    }else{
+      revert();
+    }
   }
 
   function stakeFor(address user, uint256 amount, bytes data) public onlyWhenUnlocked {
-    require(amount > 0);
-
-    // transfer tokens to stake manager
-    require(tokenObj.transferFrom(msg.sender, address(this), amount));
-
-    // check staker is present or not
-    if (stakers[user] == 0) {
-      // actual staker cannot be on index 0
-      if (stakeList.length == 0) {
-        stakeList.push(Stake({
-          amount: 0,
-          staker: address(0x0)
-        }));
-      }
-
-      // add new stake
-      stakeList.push(Stake({
-        amount: 0,
-        staker: user
-      }));
-      stakers[user] = stakeList.length - 1;
+    require(amount > minStakeAmount); 
+
+    uint256 priority = _priority(user, amount, data);
+    // actual staker cannot be on index 0
+    if (stakersList.length == 0) {
+      stakersList.push(address(0x0));
+      stakers[address(0x0)] = Staker(0, 0, new bytes(0), State.isStaker, stakingIdCount);
+      stakingIdToAddress[stakingIdCount] = address(0x0);
+      stakingIdCount.add(1);
     }
 
-    // add amount
-    uint256 stakedAmount = totalStakedFor(user).add(amount);
-    stakeList[stakers[user]].amount = stakedAmount;
-
+    // transfer tokens to stake manager
+    require(tokenObj.transferFrom(user, address(this), amount));
+    stakers[user] = Staker(currentEpoch, amount, data, State.isStaker, stakingIdCount);
+    stakingIdToAddress[stakingIdCount] = user;
+    stakeQueue.insert(priority, stakingIdCount);
+    stakersList.push(user);
+    stakingIdCount.add(1);
+    
     // update total stake
     totalStake = totalStake.add(amount);
-
-    // broadcast event
-    emit Staked(user, amount, stakedAmount, data);
+    emit Staked(user, amount, totalStake, data);
   }
+  
+  // returns validators
+  function getValidators() public returns (address[]) {
+    // add condition for currentSize of PQ
```

Comment:
Get validators should be `view`. Make necessary changes for the same.

Reply:
No modifiers? Any can call this method?

Reply:
it is updating the state of contract.